src/ptdf_calculations.jl

```diff
@@ -22,74 +22,35 @@ function binfo_check(binfo::Int)
-    linecount = length(branches)
+# version 1: use same structure as already present ###########################
 
-    bus_lookup = _make_ax_ref(nodes)
-
-    A = zeros(Float64, buscount, linecount)
-    inv_X = zeros(Float64, linecount, linecount)
-
-    #build incidence matrix
-    for (ix, b) in enumerate(branches)
-        if isa(b, PSY.DCBranch)
-            @warn("PTDF construction ignores DC-Lines")
-            continue
-        end
-
-        (fr_b, to_b) = get_bus_indices(b, bus_lookup)
-        A[fr_b, ix] = 1
-        A[to_b, ix] = -1
-
-        inv_X[ix, ix] = PSY.get_series_susceptance(b)
+function _buildptdf(branches, nodes::Vector{PSY.Bus},
+    dist_slack::Vector{Float64}, linear_solver::String = "Dense")
+    if linear_solver == "KLU"
+        ptdf, a = calculate_PTDF_matrix_KLU(branches, nodes, dist_slack)
+    elseif linear_solver == "Dense"
+        ptdf, a = calculate_PTDF_matrix_DENSE(branches, nodes, dist_slack)
+    elseif linear_solver == "MKLPardiso"
+        error("MKLPardiso still to be implemented")
     end
 
-    slacks =
-        [bus_lookup[PSY.get_number(n)] for n in nodes if PSY.get_bustype(n) == BusTypes.REF]
-    isempty(slacks) && error("System must have a reference bus!")
+    return ptdf, a
+end
 
-    slack_position = slacks[1]
-    B = gemm(
-        'N',
-        'T',
-        gemm('N', 'N', A[setdiff(1:end, slack_position), 1:end], inv_X),
-        A[setdiff(1:end, slack_position), 1:end],
-    )
-    if dist_slack[1] == 0.1 && length(dist_slack) == 1
-        (B, bipiv, binfo) = getrf!(B)
-        binfo_check(binfo)
-        S = gemm(
-            'N',
-            'N',
-            gemm('N', 'T', inv_X, A[setdiff(1:end, slack_position), :]),
-            getri!(B, bipiv),
-        )
-        @views S =
-            hcat(S[:, 1:(slack_position - 1)], zeros(linecount), S[:, slack_position:end])
-    elseif dist_slack[1] != 0.1 && length(dist_slack) == buscount
-        @info "Distributed bus"
-        (B, bipiv, binfo) = getrf!(B)
-        binfo_check(binfo)
-        S = gemm(
-            'N',
-            'N',
-            gemm('N', 'T', inv_X, A[setdiff(1:end, slack_position), :]),
-            getri!(B, bipiv),
-        )
-        @views S =
-            hcat(S[:, 1:(slack_position - 1)], zeros(linecount), S[:, slack_position:end])
-        slack_array = dist_slack / sum(dist_slack)
-        slack_array = reshape(slack_array, buscount, 1)
-        S = S - gemm('N', 'N', gemm('N', 'N', S, slack_array), ones(1, buscount))
-    elseif length(slack_position) == 0
-        @warn("Slack bus not identified in the Bus/Nodes list, can't build PTDF")
-        S = Array{Float64, 2}(undef, linecount, buscount)
-    else
-        @assert false
+function _buildptdf(A::SparseArrays.SparseMatrixCSC{Int8, Int32},
+    BA::SparseArrays.SparseMatrixCSC{T, Int32}
+    where {T <: Union{Float32, Float64}},
+    dist_slack::Vector{Float64},
+    linear_solver::String)
+    if linear_solver == "KLU"
```

Comment:
we don't need A here.

Reply:
we agreed to keep them both for the momen